rclcpp/src/rclcpp/time_source.cpp

```diff
@@ -0,0 +1,223 @@
+  if (result != associated_clocks_.end()) {
+    associated_clocks_.erase(result);
+  } else {
+    RCUTILS_LOG_ERROR("Failed to remove clock");
+  }
+}
+
+TimeSource::~TimeSource()
+{
+  if (node_) {
+    this->detachNode();
+  }
+}
+
+void TimeSource::setClock(const builtin_interfaces::msg::Time::SharedPtr msg,
+  std::shared_ptr<rclcpp::Clock> clock)
+{
+  // TODO(tfoote) Move enabledROS time logic here.
+
+  rclcpp::Time msg_time = rclcpp::Time(*msg);
+  rclcpp::Time now = clock->now();
+  auto diff = now - msg_time;
+  rclcpp::TimeJump jump;
+  jump.delta_.nanoseconds = diff.nanoseconds();
+  // TODO(tfoote) fill in time source change on jump
+
+
+  // TODO(tfoote) potential race condition with target going out of scope
+  auto active_callbacks = clock->get_triggered_callbacks(jump);
+  clock->invoke_prejump_callbacks(active_callbacks);
+  auto ret = rcl_set_ros_time_override(&(clock->rcl_clock_), msg_time.nanoseconds());
+  if (ret != RCL_RET_OK) {
+    rclcpp::exceptions::throw_from_rcl_error(
+      ret, "Failed to set ros_time_override_status");
+  }
+  clock->invoke_postjump_callbacks(active_callbacks, jump);
+}
+
+void TimeSource::clock_cb(const builtin_interfaces::msg::Time::SharedPtr msg)
+{
+  // RCUTILS_LOG_INFO("Got clock message");
+  if (!this->ros_time_valid_) {
+    enableROSTime();
+  }
+  // TODO(tfoote) switch this to be based on if there are clock publishers or use_sim_time
+  // TODO(tfoote) also setup disable
+
+  std::lock_guard<std::mutex> guard(clock_list_lock_);
+  for (auto it = associated_clocks_.begin(); it != associated_clocks_.end(); ++it) {
+    setClock(msg, *it);
+  }
+}
+
+
+void TimeSource::on_parameter_event(const rcl_interfaces::msg::ParameterEvent::SharedPtr event)
+{
+  for (auto & new_parameter : event->new_parameters) {
+    if (new_parameter.value.type == rclcpp::parameter::ParameterType::PARAMETER_BOOL) {
+      if (new_parameter.name == "use_sim_time") {
+        bool value = new_parameter.value.bool_value;
+        if (value) {
+          parameter_state_ = SET_TRUE;
+          enableROSTime();
+        } else {
+          parameter_state_ = SET_FALSE;
+          disableROSTime();
+        }
+      }
+    }
+  }
+  for (auto & changed_parameter : event->changed_parameters) {
+    if (changed_parameter.value.type == rclcpp::parameter::ParameterType::PARAMETER_BOOL) {
+      if (changed_parameter.name == "use_sim_time") {
+        bool value = changed_parameter.value.bool_value;
+        if (value) {
+          parameter_state_ = SET_TRUE;
+          enableROSTime();
+        } else {
+          parameter_state_ = SET_FALSE;
+          disableROSTime();
+        }
+      }
+    }
+  }
+  for (auto & deleted_parameter : event->deleted_parameters) {
+    if (deleted_parameter.name == "use_sim_time") {
+      // If the parameter is deleted mark it as unset but dont' change state.
+      parameter_state_ = UNSET;
+    }
+  }
```

Comment:
This function might be an argument for `event` having only one list of events, with each event having an action (new/change/delete). That way you could have looped once in this function rather than three times.

Reply:
Maybe just a convenience function to zip these three lists together for you with their action type.

Reply:
I don't think the datastructure necessarily needs to change, but definitely a function to both zip and filter for desired names would avoid a lot of this boilerplate.

Reply:
@wjwwood Is this the sort of thing you were thinking? https://github.com/ros2/rclcpp/commit/b57983c4f84d68d9c48d30974bbf4d3bcc31e623

Reply:
It's more sophisticated than I was thinking (with the parameter name filtering as well), but yeah that's more or less what I was thinking.

Reply:
I've isolated that and will submit it separately. I still need to add unit tests for it.